OrbitCore/LinuxPerfEventProcessor.h

```diff
@@ -0,0 +1,83 @@
+//-----------------------------------
+// Copyright Pierric Gimmig 2013-2019
+//-----------------------------------
+#pragma once
+
+#if __linux__
+
+#include <queue>
+#include "LinuxPerfEvent.h"
+#include "LinuxPerfEventVisitor.h"
+
+// A comparator used for the priority queue, such that pop/top will
+// always return the oldest event in the queue.
+class TimeStampCompare
+{
+public:
+    bool operator() (std::shared_ptr<LinuxPerfEvent> lhs, std::shared_ptr<LinuxPerfEvent> rhs)
+    {
+        return lhs->Timestamp() > rhs->Timestamp();
+    }
+};
+
+class LinuxPerfEventProcessor
+{
+public:
+    // While processing, we do not touch the events with a timestamp less 
+    // then 1/100 sec smaller than the most recent one in the queue.
+    // This way we can ensure, that all events (from different sources)
+    // are processed in the correct order.
+    uint64_t OFFSET = 10000000 /*ns*/;
```

Comment:
Where does the 1/100 sec come from, does it guarantee that events will always be processed in order?  Do you know in why we would receive events out of order?

Reply:
Actually, it looks like we are using a single event_buffer for different types of events.  Could it be that this is the cause of out order events?  Maybe if we had a single buffer per event type everything would always be in order and we wouldn't have to sort the events at all (and potentially get rid of the "OFFSET")?

Reply:
Okay let me give a short example of how the out-of-order events processing could happen.
Assume the following happens during profiling a process with PID/TID 42.
Time 0ns: TID 42 forked -> TID 13. (E1)
Time 10ns: TID 13 gets scheduled at CPU 0. (E2)

I will receive both events E1 and E2 (notice that they are of different type).
As there is no way of having events of different types (and for all cpu's) in on ring_buffer, we have the situation, that we might read E2 before we read E1. This is regardless of whether we read our buffers one after another, or all at once via. different threads. It might be simply the case that writing in the ring buffer belonging to E1 takes slightly longer then writing to the other ring buffer. Recall, that the time difference between the events might be super small. 

If we now would process E2 before E1, we would not know that TID 13 is actually a thread of our process TID 42. We would therefore, miss the scheduling event.

Thus, it is really important to put all events of all kinds into that event_buffer. Regardless of the OFFSET (I will come to this in a minute), this buffer's task is to sort all events by the timestamp and delay processing them so long, that we can sure, that we will not process a event with a higher timestamp as one that would be added slightly later. Here the visitor pattern actually helps, as we have no idea anymore, what kind of event we have in front of us, when processing them.

So finally the offset must just be large enough, such that delay processing so long, that there will be no event added (with a smaller timestamp) after we processed one with a larger timestamp.

I have to admit that this 1/100 sec. was just a hand-wavy guess of experiments and then adding a factor of 100 or 1000 ontop of what I saw. I will add code, to inform us, if that offset was sufficient and also will be fine with an offset of lets say a second. I will also rename the offset to "DELAY" at this describes it better.

Reply:
A lot of complexity is introduced to figure out the ProcessID from a ThreadID in order to filter the scheduling events related to our target process.  We are doing the same filtering on the Windows version, however, it is much easier since events know their PIDs.

I think we should record *all* scheduling events, not just the ones related to our process.  And that goes for Windows too.  It can help us understand why a thread is not getting scheduled on a core for example, and we'll get the full picture.

Doing this would eliminate the need of sorting events, would allow for a simpler code and would ensure that we have a *all* the scheduling information available.

Concerning the single buffer for different types of events, I don't think it's the optimal way to store the data.  I would strongly recommend cache-friendlier containers where all events of a same type would be contiguous in memory.  Under heavy load, there will likely be *a lot* of events generated and event tracking could very well become a bottleneck in the application.

Moreover, when we receive the events, we know their types.  We can easily pipe them to per-type containers where events will be tightly packed.  As it is now, using a single container of pointers to events, we lose that type information, we scatter object in memory and we need to go through virtual calls to process the data...

Reply:
As mentioned above, we now have system-wide and per PID tracing of scheduling events.
However, I would strongly prefer to keep that "inorder" preserving event buffer/processor. We also need this e.g. for profiling uprobes, such that we do not get out of order with uprobes and uretprobes.